src/callback.rs

```diff
@@ -7,86 +7,88 @@ use crate::exceptions::OverflowError;
+}
 
-    #[inline]
-    fn convert_result(py: Python, value: PyResult<Self::Source>) -> Self::Result {
-        match value {
-            Ok(val) => Self::convert(val, py),
-            Err(e) => {
-                e.restore(py);
-                Self::ERR_VALUE
-            }
-        }
-    }
+impl PyCallbackResult for libc::c_int {
+    const ERR_VALUE: Self = -1;
 }
 
-pub struct PyObjectCallbackConverter<T>(pub std::marker::PhantomData<T>);
+impl PyCallbackResult for ffi::Py_ssize_t {
+    const ERR_VALUE: Self = -1;
+}
+
+impl PyCallbackResult for () {
+    const ERR_VALUE: Self = ();
+}
 
-impl<T> CallbackConverter for PyObjectCallbackConverter<T>
+/// Convert the result of callback function into the appropriate return value.
+pub trait IntoPyCallbackResult<Target> {
+    fn convert(self, py: Python) -> PyResult<Target>;
+}
+
+impl<T, U> IntoPyCallbackResult<U> for PyResult<T>
 where
-    T: IntoPy<PyObject>,
+    T: IntoPyCallbackResult<U>,
 {
-    type Source = T;
-    type Result = *mut ffi::PyObject;
-    const ERR_VALUE: Self::Result = ptr::null_mut();
-
-    fn convert(s: Self::Source, py: Python) -> Self::Result {
-        s.into_py(py).into_ptr()
+    fn convert(self, py: Python) -> PyResult<U> {
+        self.and_then(|t| t.convert(py))
     }
 }
 
-pub struct BoolCallbackConverter;
-
-impl CallbackConverter for BoolCallbackConverter {
-    type Source = bool;
-    type Result = c_int;
-    const ERR_VALUE: Self::Result = -1;
-
-    #[inline]
-    fn convert(s: Self::Source, _py: Python) -> Self::Result {
-        s as c_int
+impl<T> IntoPyCallbackResult<*mut ffi::PyObject> for T
+where
+    T: IntoPy<PyObject>,
+{
+    fn convert(self, py: Python) -> PyResult<*mut ffi::PyObject> {
+        Ok(self.into_py(py).into_ptr())
     }
 }
 
-pub struct LenResultConverter;
+impl IntoPyCallbackResult<Self> for *mut ffi::PyObject {
+    fn convert(self, _: Python) -> PyResult<Self> {
+        Ok(self)
+    }
+}
 
-impl CallbackConverter for LenResultConverter {
-    type Source = usize;
-    type Result = isize;
-    const ERR_VALUE: Self::Result = -1;
+impl IntoPyCallbackResult<libc::c_int> for () {
+    fn convert(self, _: Python) -> PyResult<libc::c_int> {
+        Ok(0)
+    }
+}
 
-    fn convert(val: Self::Source, py: Python) -> Self::Result {
-        if val <= (isize::MAX as usize) {
-            val as isize
-        } else {
-            OverflowError::py_err(()).restore(py);
-            -1
-        }
+impl IntoPyCallbackResult<libc::c_int> for bool {
+    fn convert(self, _: Python) -> PyResult<libc::c_int> {
+        Ok(self as c_int)
     }
 }
 
-pub struct UnitCallbackConverter;
+impl IntoPyCallbackResult<()> for () {
+    fn convert(self, _: Python) -> PyResult<()> {
+        Ok(())
+    }
+}
 
-impl CallbackConverter for UnitCallbackConverter {
-    type Source = ();
-    type Result = c_int;
-    const ERR_VALUE: Self::Result = -1;
+pub struct LenResultConverter(pub usize);
```

Comment:
Now, this is rather a `Result` than a converter.
How about renaming it `LenResult`. `CbLenResult` or so?

Reply:
I don't like using `Result` when it's not an alias of `std::result::Result` so I'm going to call it `LenCallbackOutput`

Reply:
And in fact I'm going to rename the `PyCallbackResult` trait to `PyCallbackOutput` in the same way.